### ISSUE TITLE
Merchant should be able to enable out-of-stock-ordering for each combination

### ISSUE THREAD

**Issue by Hlavtox** [PrestaShop/PrestaShop#24279]
#### Is your feature request related to a problem?
Related to my previous issue: https://github.com/PrestaShop/PrestaShop/issues/20536
I cannot set, if the system should allow ordering out of stock items for different combinations of products. 

My clients are pushing on me because customers are ordering things that are not in stock => angry customers, when you tell them that the product they ordered will take X weeks to deliver.

#### Use case
I am selling a t-shirt with three combinations - S, M, L.
- I have only S size in stock.
- The supplier has only M in stock. - I want the user to be able to order this.
- The supplier says L will take month to deliver, because they don't have it in stock. - I know there are problems with delivery of products, which the supplier does not have in stock. I don't want the user to be able to order this.

#### Describe the solution you'd like
Ability to set out of stock behavior for each combination.

#### Screenshot
![comb](https://user-images.githubusercontent.com/6097524/116679351-b4ee9080-a9aa-11eb-99ab-7ca564f42c11.JPG)

**Comment by florine2623** [comment]
Hello @Hlavtox ,

Thank you for your suggestion. The Product Team will take it into consideration for future developments. 

Please be aware that there is no guarantee that this feature will be developed anytime soon. If this is important to you, we strongly suggest you consider [hiring a specialist](https://www.prestashop.com/en/experts) to help you create it. If you are willing to implement this feature into PrestaShop as a [Pull Request](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/), the teams will be happy to help you by ensuring that it is appropriate and coherent with the rest of the project. If that is the case, please let us know!

Thank you

**Comment by ghost** [comment]
Hey @Hlavtox 

I made this feature on a shop in 1.6 it's very usefull 

On an other shop in 177 i use a module and it's work good, but it's a real improvment for prestashop i agree !

![Capture d’écran de 2021-05-06 12-21-08](https://user-images.githubusercontent.com/5064590/117283525-27a5b300-ae66-11eb-85ca-62591f41a6a3.png)

**Comment by jf-viguier** [comment]
A must have

**Comment by jolelievre** [comment]
We tried to integrate this feature in v8.0 sadly it generates too many impact and we don't have time to stabilize it correctly considering how close we are to the feature freeze This will be handled in a future version I level here the PR that started the integration https://github.com/PrestaShop/PrestaShop/pull/28589

**Comment by korsgaard** [comment]
Yes - please! - this makes so much sense also with the new backordering labels foreach combination in 8.1!

**Comment by mauweb** [comment]
> EHI@Hlavtox
> 
> Ho creato questa funzionalità in un negozio nella versione 1.6, è molto utile
> 
> In un altro negozio a 177 utilizzo un modulo e funziona bene, ma è un vero miglioramento per PrestaShop, sono d'accordo!
> 
> ![Cattura d'écran del 2021-05-06 12-21-08](https://user-images.githubusercontent.com/5064590/117283525-27a5b300-ae66-11eb-85ca-62591f41a6a3.png)

Which module do you use for PS 1.7?

> Hey @Hlavtox
> 
> I made this feature on a shop in 1.6 it's very usefull
> 
> On an other shop in 177 i use a module and it's work good, but it's a real improvment for prestashop i agree !
> 
> ![Capture d’écran de 2021-05-06 12-21-08](https://user-images.githubusercontent.com/5064590/117283525-27a5b300-ae66-11eb-85ca-62591f41a6a3.png)
-----------------------------------------------------------------


Which module do you use for PS 1.7?

**Comment by SymeR17** [comment]
Hello,
Has anyone managed to resolve this issue? It's still present.
Have you found a module ?

**Comment by boo-code** [comment]
Measured the current state, because the data model is closer to this than the discussion suggests and the blocker is narrower than "not supported".

**The column already exists per combination.** `stock_available` carries `out_of_stock` on every row, including the per-combination ones. On a stock 9.2 install a product with four combinations has five rows, each with its own value:

```
id_product  id_product_attribute  quantity  out_of_stock
2           0                     2099      2
2           9                     1199      2
2           10                    300       2
2           11                    300       2
2           12                    300       2
```

They are all equal only because nothing ever lets them diverge.

**Two pieces keep them in lockstep.** `StockAvailable::outOfStock()` hardcodes `id_product_attribute = 0`, so it always answers the product-level value - and its twelve call sites include the one that decides this feature, `Cart::updateQty()` at the add-to-cart gate, which has `$id_product_attribute` in hand and does not pass it. And `Combination.php:217` copies the product-level value down into each new combination row, while `StockAvailable::setQuantity()` seeds new rows the same way.

So the work is not "store a value per combination" - that already happens - it is to stop discarding it on read and stop overwriting it on write.

**@korsgaard's point is worth sharpening**, because it makes the current state internally inconsistent rather than merely limited: the *labels* already are per combination. `product_attribute_lang` has `available_now` and `available_later`, and `ProductLazyArray` prefers `$combinationData['available_later']` over the product's. So a shop can already tell a customer something different per size, while the rule that decides whether that size can be ordered at all stays product-wide.

**On feasibility**, @jolelievre's #28589 is the measurement that matters: 29 files, including `Cart.php`, `Combination.php`, `Product.php`, `StockAvailable.php`, all four order command and query handlers, and the combination CQRS stack. That matches the call-site list above, and it is why "too many impacts to stabilize" was the right call in a feature freeze rather than a reason the feature is wrong.

No patch prepared, deliberately: #28589 already exists as the reference implementation and re-doing it would be a rival rather than a contribution. What would unblock it is a decision on one semantic question the code cannot answer by itself - whether a combination's value is an override with an explicit "inherit from product" state, or a plain per-combination value seeded once at creation. The first keeps `setQuantity()`'s current seeding meaningful, the second makes it a bug; that choice determines how much of those 29 files actually has to change.
